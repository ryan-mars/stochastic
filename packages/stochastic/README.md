# Flight Ops Example

Example service responsible for flight operations at a hypothetical airline.

## List of things we need (in small chunks)

- [ ] Figure out how we can give NodejsFunction the entry (file path) and handler (exported function name)
- [ ] POC of not having the entire graph defined in one file
- [ ] POC of handler -> NodejsFunction
- [ ] POC of passing dependent clients to a handler function
- [ ] Use SSM instead of environment variables for Lambda Function dependencies
- [ ] Command Handler Lambda or Ephemeral Command Handler?
- [ ] Pseudo code a CDK stack that uses (synthesizes?) an even storm in the stack constructor
- [ ] A CDK construct that takes an Event Storm and generates all of the actual CDK resources.
- [ ] Event Storm maintains a graph
- [ ] POC of working policy
- [ ] Store a domain event in DynamoDB
- [ ] Forward domain events to an SNS topic
- [ ] Subscribe event-handlers (policy, read-model updater) to filtered topic queue 
- [ ] Metabolism name?
- [ ] Identify certain events as public / can be subscribed to by outside services
- [ ] Publish public events to EventBridge
- [ ] Subscribe policy to public event from another Event Storm (using Event Bridge)
- [ ] Add a HTTP API to an event storm
- [ ] Add a GraphQL API to an event storm
- [ ] Modify CDK resources created by a synthesized Event Storm (aspect oriented?)

## Design Concepts

- Event Storm and it’s collaborators (Aggregate, Policy, Event, ReadModel, etc…) have zero CDK code in them (but may leverage types for storing configuration?)
- An Event Storm “synthesizer” CDK Construct takes an Event Storm graph and creates the necessary infrastructure
