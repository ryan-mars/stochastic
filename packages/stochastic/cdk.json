{
  "app": "npx ts-node src/flight-stack.ts"
}
