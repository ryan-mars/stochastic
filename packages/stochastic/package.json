{
  "name": "stochastic",
  "private": true,
  "version": "0.1.0",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "scripts": {
    "clean": "rm -rf lib",
    "build": "tsc",
    "watch": "tsc -w",
    "cdk": "cdk",
    "test": "jest --passWithNoTests"
  },
  "keywords": [],
  "author": "Ryan Marsh",
  "devDependencies": {
    "@types/jest": "^26.0.23",
    "@types/node": "^14.14.35",
    "jest": "^26.6.3",
    "ts-jest": "^26.5.6",
    "ts-node": "^9.1.1",
    "typescript": "^4.2.3"
  },
  "peerDependencies": {
    "superstruct": "0.14.2"
  },
  "dependencies": {
    "ksuid": "^2.0.0",
    "superstruct": "0.14.2"
  }
}
