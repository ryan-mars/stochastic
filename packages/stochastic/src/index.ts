export * from "./aggregate";
export * from "./command";
export * from "./component";
export * from "./dependencies";
export * from "./event-storm";
export * from "./event";
export * from "./policy";
export * from "./query";
export * from "./read-model";
export * from "./runtime";
export * from "./type";
