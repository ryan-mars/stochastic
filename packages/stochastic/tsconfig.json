{
  "compilerOptions": {
    "outDir": "lib/cjs",
    "rootDir": "src",
    "typeRoots": ["../../../node_modules/@types", "node_modules/@types"]
  },
  "include": ["src"],
  "extends": "./tsconfig.base.json"
}
