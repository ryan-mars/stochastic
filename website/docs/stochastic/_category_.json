{
  "label": "Stochastic Framework",
  "position": 3
}
