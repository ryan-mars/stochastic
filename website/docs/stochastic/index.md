---
title: "Overview"
sidebar_position: 1
slug: /stochastic/
---

# Stochastic

This section should help developers learn how to be productive with Stochastic by translating event storm models to working software.

It should include:

- Quick overview
- Sample app guide
- How to guides for specific tasks
- Design principles
- TypeScript API docs
- Deployment instructions
