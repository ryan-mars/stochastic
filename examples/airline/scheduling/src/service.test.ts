import { DomainEventEnvelope } from "stochastic";
import {
  AddScheduledFlightCommandHandler,
  AddScheduledFlightIntent,
  CreateFlightCommandHandler,
  CreateFlightIntent,
  FlightCreatedEvent,
  FlightSchedule,
  FlightScheduleAggregate,
  ScheduledFlightAdded,
} from "./service";

describe("Scheduling service", () => {
  const flightCreated = new DomainEventEnvelope({
    id: "1sC26Tx3VUi42mghcNopBYsRxD9",
    time: new Date("2021-05-07T04:34:35.302Z"),
    source: "FlightScheduleAggregate",
    source_id: "PA576",
    payload: new FlightCreatedEvent({
      flightNo: "PA576",
      aircraftType: "B787-9",
      origin: "SFO",
      destination: "MIA",
    }),
  });
  const scheduledFlightAdded = new DomainEventEnvelope({
    id: "1sC539ZsfhN9bHV4K8jZqtSz9bN",
    time: new Date("2021-05-07T04:58:48.459Z"),
    source: "FlightScheduleAggregate",
    source_id: "PA576",
    payload: new ScheduledFlightAdded({
      flightNo: "PA576",
      add: {
        day: "2021-06-11",
        scheduledArrival: new Date("2021-06-12T02:30:00.000Z"),
        scheduledDeparture: new Date("2021-06-11T20:30:00.000Z"),
      },
    }),
  });
  it("should reduce aggregate state from events", () => {
    let state = FlightScheduleAggregate.reducer(FlightScheduleAggregate.initialState(), flightCreated.payload);
    state = FlightScheduleAggregate.reducer(state, scheduledFlightAdded.payload);
    expect(state).toMatchSnapshot();
  });

  const undefinedAggregate = {
    get: async (key: string) => ({ state: FlightScheduleAggregate.initialState(), events: [] }),
  };

  it("creates a flight", async () => {
    const event = await CreateFlightCommandHandler.execute(
      new CreateFlightIntent({
        flightNo: "PA576",
        aircraftType: "B787-9",
        origin: "SFO",
        destination: "MIA",
      }),
      undefinedAggregate,
    );
    expect(event).toMatchSnapshot();
  });

  it("adds a scheduled flight", async () => {
    const event = await AddScheduledFlightCommandHandler.execute(
      new AddScheduledFlightIntent({
        flightNo: "PA576",
        add: {
          scheduledDeparture: new Date("2021-06-11T12:30:00-08:00"),
          scheduledArrival: new Date("2021-06-11T21:30:00-05:00"),
          day: "2021-06-11",
        },
      }),
      {
        get: async (key) => ({
          state: new FlightSchedule({
            flightNo: "PA576",
            aircraftType: "B787-9",
            origin: "SFO",
            destination: "MIA",
            days: new Map(),
          }),
          events: [],
        }),
      },
    );
    expect(event).toMatchSnapshot();
  });
});
