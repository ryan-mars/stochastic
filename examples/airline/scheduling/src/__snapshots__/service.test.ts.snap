// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scheduling service adds a scheduled flight 1`] = `
Array [
  ScheduledFlightsAdded {
    "__typename": "ScheduledFlightsAdded",
    "flights": Array [
      Object {
        "arrivalTime": "928p",
        "day": "2021-06-11",
        "departureTime": "1210p",
        "flightNo": "PA576",
      },
      Object {
        "arrivalTime": "502p",
        "day": "2021-06-11",
        "departureTime": "700a",
        "flightNo": "PA872",
      },
    ],
    "route": "SFO-MIA",
  },
]
`;

exports[`Scheduling service adds routes 1`] = `
Array [
  RouteAdded {
    "__typename": "RouteAdded",
    "route": "SFO-MIA",
  },
]
`;

exports[`Scheduling service reduces state from events 1`] = `
RouteSchedule {
  "__typename": "RouteSchedule",
  "flights": Array [
    Object {
      "arrivalTime": "928p",
      "day": "2021-06-11",
      "departureTime": "1210p",
      "flightNo": "PA576",
    },
    Object {
      "arrivalTime": "502p",
      "day": "2021-06-11",
      "departureTime": "700a",
      "flightNo": "PA872",
    },
    Object {
      "arrivalTime": "513p",
      "day": "2021-06-11",
      "departureTime": "700a",
      "flightNo": "PA738",
    },
    Object {
      "arrivalTime": "928p",
      "day": "2021-06-12",
      "departureTime": "1210p",
      "flightNo": "PA576",
    },
    Object {
      "arrivalTime": "502p",
      "day": "2021-06-12",
      "departureTime": "700a",
      "flightNo": "PA872",
    },
    Object {
      "arrivalTime": "513p",
      "day": "2021-06-12",
      "departureTime": "700a",
      "flightNo": "PA738",
    },
  ],
  "route": "SFO-MIA",
}
`;
